tst/testinstall/kernel/hpc/threadapi.tst

```diff
@@ -0,0 +1,397 @@
+#
+# Tests for functions defined in src/threadapi.c
+#
+#@local t, t2, region, ch, f, sem, b, sv, old_state, tmp, tmp2
+#
+#@if IsHPCGAP
+gap> START_TEST("kernel/threadapi.tst");
+
+#
+# threads
+#
+gap> CreateThread();
+Error, CreateThread: Needs at least one function argument
+gap> CreateThread(1);
+Error, CreateThread: Needs at least one function argument
+gap> CreateThread(x->x);
+Error, CreateThread: <func> expects 1 arguments, but got 0
+gap> WaitThread(fail);
+Error, WaitThread: <thread> must be a thread object (not the value 'fail')
+gap> KillThread(fail);
+Error, KillThread: <thread> must be a thread object or an integer between 0 an\
+d MAX_THREADS - 1 (not the value 'fail')
+gap> InterruptThread(fail, fail);
+Error, InterruptThread: <thread> must be a thread object or an integer between\
+ 0 and MAX_THREADS - 1 (not the value 'fail')
+
+#
+gap> SetInterruptHandler(fail, fail);
+Error, SetInterruptHandler: <handler> must be an integer between 0 and 100 (no\
+t the value 'fail')
+gap> SetInterruptHandler(0, 0);
+Error, SetInterruptHandler: <func> must be a parameterless function or 'fail' \
+(not the integer 0)
+gap> SetInterruptHandler(0, {a,b}->fail);
+Error, SetInterruptHandler: <func> must be a parameterless function or 'fail' \
+(not a function)
+
+#
+gap> PauseThread(fail);
+Error, PauseThread: <thread> must be a thread object or an integer between 0 a\
+nd MAX_THREADS - 1 (not the value 'fail')
+gap> ResumeThread(fail);
+Error, ResumeThread: <thread> must be a thread object or an integer between 0 \
+and MAX_THREADS - 1 (not the value 'fail')
+
+#
+gap> t := CreateThread(function() MicroSleep(100); end);;
+gap> PauseThread(t);
+gap> ResumeThread(t);
+gap> WaitThread(t); # FIXME: sometimes deadlocks
```

Comment:
@rbehrends it seems `WaitThread` sometimes (???) ends up blocking the calling thread if the thread `t` has terminated before. Is that right and intentional? If so, how can one use `WaitThread` safely and without a race condition? Actually, assuming that thread ids are reused, I don't see how one can use this safely at all unless every thread one wants to wait for never "just exits", but instead uses careful control logic (e.g. a semaphore or so) to block its exit until it receives "permission to die" from its controller. Whereas for pthreads, to my understanding this problem does not exist, because each `pthread_t` is unique and cannot be reused until explicitly reclaimed via `pthread_destroy`. Well, there is also `pthread_detach` but HPC-GAP doesn't allow calling that.

Am I missing something? What?

The one way to solve this I can think of is to *require* use of thread objects (drop support for specifying threads via a raw thread ID, which IMHO is problematic anyway). Then, `FuncWaitThread` could pass a given `T_THREAD` object `obj` to `JoinThread`, which then could validate `obj` by comparing it against `thread_data[ThreadID(obj)].thread_object`.

Reply:
There shouldn't be a problem here. Note that while there may be multiple thread objects that share the same thread ID, at most one will not have `THREAD_JOINED` set. This means that `JoinThread()` will never be called more than once for the thread associated with a given thread object.

Reply:
I'll add that I am wondering if it's actually the `PauseThread()` and `ResumeThread()` that directly or indirectly cause `WaitThread()` to block. They do manipulate the exec table in `SetInterrupt()` in particular, and a bug there could disrupt subsequent thread execution.